services/src/main/java/org/keycloak/protocol/oid4vc/issuance/credentialbuilder/SdJwtCredentialBuilder.java

```diff
@@ -0,0 +1,99 @@
+ */
+
+package org.keycloak.protocol.oid4vc.issuance.credentialbuilder;
+
+import org.keycloak.protocol.oid4vc.model.CredentialBuildConfig;
+import org.keycloak.protocol.oid4vc.model.CredentialSubject;
+import org.keycloak.protocol.oid4vc.model.Format;
+import org.keycloak.protocol.oid4vc.model.VerifiableCredential;
+import org.keycloak.sdjwt.DisclosureSpec;
+import org.keycloak.sdjwt.SdJwt;
+import org.keycloak.sdjwt.SdJwtUtils;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+public class SdJwtCredentialBuilder implements CredentialBuilder {
+
+    public static final String ISSUER_CLAIM = "iss";
+    public static final String VERIFIABLE_CREDENTIAL_TYPE_CLAIM = "vct";
+    public static final String CREDENTIAL_ID_CLAIM = "jti";
+
+    private final String credentialIssuer;
+
+    public SdJwtCredentialBuilder(String credentialIssuer) {
+        this.credentialIssuer = credentialIssuer;
+    }
+
+    @Override
+    public String getSupportedFormat() {
+        return Format.SD_JWT_VC;
+    }
+
+    @Override
+    public SdJwtCredentialBody buildCredentialBody(
+            VerifiableCredential verifiableCredential,
+            CredentialBuildConfig credentialBuildConfig
+    ) throws CredentialBuilderException {
+        // Retrieve claims
+        CredentialSubject credentialSubject = verifiableCredential.getCredentialSubject();
+        Map<String, Object> claimSet = credentialSubject.getClaims();
+
+        // Put all claims into the disclosure spec, except the one to be kept visible
+        DisclosureSpec.Builder disclosureSpecBuilder = DisclosureSpec.builder();
+        claimSet.entrySet()
+                .stream()
+                .filter(entry -> !credentialBuildConfig.getVisibleClaims().contains(entry.getKey()))
+                .forEach(entry -> {
+                    if (entry instanceof List<?> listValue) {
+                        // FIXME: Unreachable branch. The intent was probably to check `entry.getValue()`,
+                        //  but changing just that will expose the array field name and break many tests.
+                        //  Needs further discussion on the wanted behavior.
+
+                        IntStream.range(0, listValue.size())
+                                .forEach(i -> disclosureSpecBuilder
+                                        .withUndisclosedArrayElt(entry.getKey(), i, SdJwtUtils.randomSalt())
+                                );
+                    } else {
+                        disclosureSpecBuilder.withUndisclosedClaim(entry.getKey(), SdJwtUtils.randomSalt());
+                    }
+                });
+
+        // Populate configured fields (necessarily visible)
+        claimSet.put(ISSUER_CLAIM, credentialIssuer);
+        claimSet.put(VERIFIABLE_CREDENTIAL_TYPE_CLAIM, credentialBuildConfig.getCredentialType());
+        claimSet.put(CREDENTIAL_ID_CLAIM, CredentialBuilderUtils.createCredentialId(verifiableCredential));
```

Comment:
I have the impression that `jti` shouldn't be included. For sure, [SD-JWT-VC](https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-08.html#name-registered-jwt-claims) doesn't requires it (or mentioned it).

IMHO, it must not be included, given that - in theory - `jti` uniquely identifies a credential instance (issued from a given issuer) and can be used to track the presentations of such a SD-JWT

Reply:
Thank you for taking the time to leave a comment despite your heavy schedule. I concur that including an identifier might indeed be a source of correlation across presentations. Since this is just a refactored code, let me tag @francis-pouatcha to confirm it is safe to simply remove the line.

Reply:
@babisRoutis Dropping the JIT doesn't necessarily increase security. Reusing an SD-JWT allows an attacker to correlate requests by simply hashing the SD-JWT.

Reply:
@babisRoutis In the original code, I removed ```nbf``` and ```iat``` to reduce correlation factors.

```java
        // nbf, iat and exp are all optional. So need to be set by a protocol mapper if needed
        // see: https://www.ietf.org/archive/id/draft-ietf-oauth-sd-jwt-vc-03.html#name-registered-jwt-claims

        // Use vct as type for sd-jwt.
        rootNode.put(VERIFIABLE_CREDENTIAL_TYPE_CLAIM, vct.getValue());
        rootNode.put(CREDENTIAL_ID_CLAIM, JwtSigningService.createCredentialId(verifiableCredential));
```

In a future version, we could proceed the same, we can show it is a correlation factor.

Reply:
@babisRoutis The absence of `nonce` and `iat` significantly amplifies the importance of `jti` in mitigating replay attacks, especially for non-holder-bound tokens (tokens that are not tied to a specific user or device).

While the entire token itself could be used as an identifier, the community and implementations commonly use `jti` for revocation management and replay prevention.

Making it optional would definitely require a comprehensive "sensibilization campaign" to ensure developers understand the associated risks and implement appropriate safeguards.

Reply:
**Suggestion**: If there are no further comments or objections by tomorrow at 23:59 UTC, I suggest @IngridPuppet  proceeds with removing the `jti` and we give it the same optional status as `iat`.

Reply:
Hi @francis-pouatcha 

Thank you for considering my comment and I agree to your suggestion to treat `jti` as optional.

PS: Have no objection to your arguments about `jti`, especially in the case that a kb-jwt is not required.

Reply:
From the breakout session of 12-18-2024, @IngridPuppet will add a mapper for jti and make it optional that way.

Reply:
I implemented the agreed-upon change. With the mapper solution, disclosing or not the `jti` field to inject is also configurable. Please could you check ? @babisRoutis @francis-pouatcha